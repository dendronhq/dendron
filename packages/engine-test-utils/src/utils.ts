import {
  CONSTANTS,
  DendronConfig,
  VaultUtils,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings,
} from "@dendronhq/common-all";
import { readYAML } from "@dendronhq/common-server";
import { AssertUtils } from "@dendronhq/common-test-utils";
import { DConfig, Git } from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";

export async function checkString(body: string, ...match: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      match,
    })
  ).toBeTruthy();
}

export async function checkNotInString(body: string, ...nomatch: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      nomatch,
    })
  ).toBeTruthy();
}

const getWorkspaceFolders = (wsRoot: string) => {
  const wsPath = path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME);
  const settings = fs.readJSONSync(wsPath) as WorkspaceSettings;
  return _.toArray(settings.folders);
};

export function checkVaults(opts: WorkspaceOpts, expect: any) {
  const { wsRoot, vaults } = opts;
  const configPath = DConfig.configPath(opts.wsRoot);
  const config = readYAML(configPath) as DendronConfig;
  expect(_.sortBy(config.vaults, ["fsPath", "workspace"])).toEqual(
    _.sortBy(vaults, ["fsPath", "workspace"])
  );
  const wsFolders = getWorkspaceFolders(wsRoot);
  expect(wsFolders).toEqual(
    vaults.map((ent) => {
      const out: WorkspaceFolderRaw = { path: VaultUtils.getRelPath(ent) };
      if (ent.name) {
        out.name = ent.name;
      }
      return out;
    })
  );
}

export class GitTestUtils {
  static async createRepoForWorkspace(wsRoot: string) {
    const git = new Git({ localUrl: wsRoot });
    await git.init();
    await git.add("dendron.yml");
    await git.commit({ msg: "init" });
  }

  /** Creates a "bare" git repository, to be used as the remote for a workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteCreate(remoteDir: string) {
    const git = new Git({ localUrl: remoteDir, bare: true });
    await git.init();
  }

  /** Adds a bare repository created with `createRemote` as the remote for the workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteAdd(wsRoot: string, remoteDir: string) {
    const git = new Git({ localUrl: wsRoot, remoteUrl: remoteDir });
    await git.remoteAdd();
    // Need to push to be able to set up remote tracking branch
    await git.push({ remote: "origin", branch: await git.getCurrentBranch() });
  }

  /** Set up a workspace with a remote, intended to be used when testing pull or push functionality.
   *
   * @param wsRoot Directory where the workspace will be stored.
   * @param remoteDir Directory where the remote will be stored. The workspace will pull and push to this remote.
   */
  static async createRepoForRemoteWorkspace(wsRoot: string, remoteDir: string) {
    await this.createRepoForWorkspace(wsRoot);
    await this.remoteCreate(remoteDir);
    await this.remoteAdd(wsRoot, remoteDir);
  }

  static async createRepoWithReadme(root: string) {
    const git = new Git({ localUrl: root });
    await git.init();
    const readmePath = path.join(root, "README.md");
    fs.ensureFileSync(readmePath);
    await git.add(".");
    await git.commit({ msg: "init" });
  }
}
