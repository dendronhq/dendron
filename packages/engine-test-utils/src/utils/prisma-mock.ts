// import { PrismaClient } from "@prisma/client";
// import { mockDeep, mockReset, DeepMockProxy } from "jest-mock-extended";
// import { dendronPrisma } from "@dendronhq/engine-server";

// jest.mock("./client", () => ({
//   __esModule: true,
//   default: mockDeep<PrismaClient>(),
// }));

// beforeEach(() => {
//   mockReset(prismaMock);
// });

// export const prismaMock =
//   dendronPrisma as unknown as DeepMockProxy<PrismaClient>;
