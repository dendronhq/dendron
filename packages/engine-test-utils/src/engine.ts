import {
  CleanDendronSiteConfig,
  CONSTANTS, DendronConfig, DEngineClient,
  DVault,
  DWorkspace,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings
} from "@dendronhq/common-all";
import {
  getDurationMilliseconds,
  tmpDir,
  vault2Path
} from "@dendronhq/common-server";
import {
  GenTestResults,
  PostSetupHookFunction,
  PreSetupHookFunction,
  RunEngineTestFunctionOpts,
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
  SetupTestFunctionV4,
  sinon,
  TestResult
} from "@dendronhq/common-test-utils";
import { LaunchEngineServerCommand } from "@dendronhq/dendron-cli";
import {
  createEngine as engineServerCreateEngine,
  DConfig,
  WorkspaceService
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import os from "os";
import path from "path";
import { SinonStub } from "sinon";
import { ENGINE_HOOKS } from "./presets";
import { GitTestUtils } from "./utils";

export type TestSetupWorkspaceOpts = {
  /**
   * Vaults to initialize engine with
   * Defaults to following if not set
   * [
   *    { fsPath: "vault1" },
   *    { fsPath: "vault2" },
   *    { fsPath: "vault3", name: "vaultThree" },
   *  ]
   */
  vaults?: DVault[];
  /**
   * Modify dendron config before initialization
   */
  modConfigCb?: (config: DendronConfig) => DendronConfig;
};

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<{engine: DEngineClient, port?: number}>;

/**
 * Create an {@link DendronEngine}
 */
export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return {engine: engineServerCreateEngine(opts) as DEngineClient, port: undefined};
}

export { DEngineClient, DVault, WorkspaceOpts };

/**
 * Create a server
 * @param opts
 * @returns
 */
// @ts-ignore
export async function createServer(opts: WorkspaceOpts) {
  return await new LaunchEngineServerCommand().enrichArgs({
    wsRoot: opts.wsRoot,
  });
}

/**
 * Create an {@link DendronEngineClient}
 */
export async function createEngineFromServer(
  opts: WorkspaceOpts
): Promise<any> {
  const { engine, port } = await createServer(opts);
  await engine.init();
  return {engine, port};
}

export function createSiteConfig(
  opts: Partial<CleanDendronSiteConfig> &
    Required<Pick<CleanDendronSiteConfig, "siteRootDir" | "siteHierarchies">>
): CleanDendronSiteConfig {
  let copts = {
    siteNotesDir: "docs",
    siteUrl: "https://localhost:8080",
    ...opts,
  };
  return {
    ...copts,
    siteIndex: DConfig.getSiteIndex(copts),
  };
}

/**
 *
 * @param opts.asRemote: add git repo
 * @param opts.wsRoot: override given wsRoot
 * @returns
 */
export async function setupWS(opts: {
  vaults: DVault[];
  workspaces?: DWorkspace[];
  asRemote?: boolean;
  wsRoot?: string;
}) {
  const wsRoot = opts.wsRoot || tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  ws.createConfig();
  const config = ws.config;
  let vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault, config, updateConfig: false });
      return vault;
    })
  );
  config.vaults = _.sortBy(config.vaults, "fsPath");
  if (config.site.duplicateNoteBehavior) {
    config.site.duplicateNoteBehavior.payload = (
      config.site.duplicateNoteBehavior.payload as string[]
    ).sort();
  }
  ws.setConfig(config);
  if (opts.workspaces) {
    const vaultsFromWs = await _.reduce(
      opts.workspaces,
      async (resp, ent) => {
        await resp;
        await WorkspaceService.createWorkspace({
          wsRoot: path.join(wsRoot, ent.name),
          vaults: ent.vaults,
        });
        return ws.addWorkspace({ workspace: ent });
      },
      Promise.resolve({ vaults: [] } as { vaults: DVault[] })
    );
    vaults = vaults.concat(vaultsFromWs.vaults);
  }
  if (opts.asRemote) {
    await GitTestUtils.createRepoWithReadme(wsRoot);
  }
  return { wsRoot, vaults };
}

export type RunEngineTestV5Opts = {
  preSetupHook?: SetupHookFunction;
  createEngine?: AsyncCreateEngineFunction;
  extra?: any;
  expect: any;
  workspaces?: DWorkspace[];
  setupOnly?: boolean;
  initGit?: boolean;
  initHooks?: boolean;
  addVSWorkspace?: boolean;
  /**
   * Path to preset wsRoot
   */
  wsRoot?: string;
} & TestSetupWorkspaceOpts;

export type RunEngineTestFunctionV5<T = any> = (
  opts: RunEngineTestFunctionOpts & { extra?: any; engineInitDuration: number, port?: number }
) => Promise<TestResult[] | void | T>;

export class TestPresetEntryV5 {
  public preSetupHook: PreSetupHookFunction;
  public postSetupHook: PostSetupHookFunction;
  public testFunc: RunEngineTestFunctionV4;
  public extraOpts: any;
  public setupTest?: SetupTestFunctionV4;
  public genTestResults?: GenTestResults;
  public vaults: DVault[];
  public workspaces: DWorkspace[];

  constructor(
    func: RunEngineTestFunctionV5,
    opts?: {
      preSetupHook?: PreSetupHookFunction;
      postSetupHook?: PostSetupHookFunction;
      extraOpts?: any;
      setupTest?: SetupTestFunctionV4;
      genTestResults?: GenTestResults;
      vaults?: DVault[];
    }
  ) {
    let {
      preSetupHook,
      postSetupHook,
      extraOpts,
      setupTest,
      genTestResults,
      workspaces,
    } = _.defaults(opts, {
      workspaces: [],
    });
    this.preSetupHook = preSetupHook ? preSetupHook : async () => {};
    this.postSetupHook = postSetupHook ? postSetupHook : async () => {};
    this.testFunc = _.bind(func, this);
    this.extraOpts = extraOpts;
    this.setupTest = setupTest;
    this.genTestResults = _.bind(
      genTestResults ? genTestResults : async () => [],
      this
    );
    this.workspaces = workspaces;
    this.vaults = opts?.vaults || [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      {
        name: "vaultThree",
        fsPath: "vault3",
      },
    ];
  }
}

/**
 *
 * To create empty workspace, initilizae with `vaults = []`
 * @param func
 * @param opts.vaults: By default, initiate 3 vaults {vault1, vault2, (vault3, "vaultThree")}
 * @param opts.preSetupHook: By default, initiate empty
 * @param opts.wsRoot: Override the randomly generated test directory for the wsRoot
 * @returns
 */
export async function runEngineTestV5(
  func: RunEngineTestFunctionV5,
  opts: RunEngineTestV5Opts
): Promise<any> {
  const {
    preSetupHook,
    extra,
    vaults: vaultsInit,
    createEngine,
    initGit,
    workspaces,
    addVSWorkspace,
  } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    // third vault has diff name
    vaults: [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ],
    addVSWorkspace: false,
  });
  try {
    // make sure tests don't overwrite local homedir contents
    TestEngineUtils.mockHomeDir();
    const { wsRoot, vaults } = await setupWS({
      vaults: vaultsInit,
      workspaces,
      wsRoot: opts.wsRoot,
    });
    if ((opts.initHooks, vaults)) {
      fs.ensureDirSync(path.join(wsRoot, CONSTANTS.DENDRON_HOOKS_BASE));
    }
    await preSetupHook({ wsRoot, vaults });
    const resp = await createEngine({ wsRoot, vaults });
    const engine = resp.engine;
    const start = process.hrtime();
    const initResp = await engine.init();
    if (addVSWorkspace) {
      fs.writeJSONSync(
        path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME),
        {
          folders: vaults.map((ent) => ({
            path: ent.fsPath,
            name: ent.name,
          })) as WorkspaceFolderRaw[],
          settings: {},
          extensions: {},
        } as WorkspaceSettings,
        { spaces: 4 }
      );
    }
    const engineInitDuration = getDurationMilliseconds(start);
    const testOpts = {
      wsRoot,
      vaults,
      engine,
      initResp,
      port: resp.port,
      extra,
      config: engine,
      engineInitDuration,
    };
    if (initGit) {
      await GitTestUtils.createRepoForWorkspace(wsRoot);
      await Promise.all(
        vaults.map((vault) => {
          return GitTestUtils.createRepoWithReadme(
            vault2Path({ vault, wsRoot })
          );
        })
      );
    }
    if (opts.setupOnly) {
      return testOpts;
    }
    const results = (await func(testOpts)) || [];
    await runJestHarnessV2(results, expect);
    return { opts: testOpts, resp: undefined, wsRoot };
  } finally {
    // restore sinon so other tests can keep running
    sinon.restore();
  }
}

export function testWithEngine(
  prompt: string,
  func: RunEngineTestFunctionV4,
  opts?: Omit<RunEngineTestV5Opts, "expect"> & { only?: boolean }
) {
  if (opts?.only) {
    return test.only(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  } else {
    return test(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  }
}

export class TestEngineUtils {
  static mockHomeDir(dir?: string): SinonStub {
    if (_.isUndefined(dir)) dir = tmpDir().name;
    return sinon.stub(os, "homedir").returns(dir);
  }

  static vault1(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault1" })!;
  }

  static vault2(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault2" })!;
  }

  static vault3(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault3" })!;
  }
}
