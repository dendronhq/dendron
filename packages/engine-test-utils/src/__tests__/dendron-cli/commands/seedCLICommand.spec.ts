import {
  DendronError,
  DEngineClient,
  DVault,
  ERROR_STATUS,
  SeedCommands,
  SeedConfig,
} from "@dendronhq/common-all";
import { tmpDir } from "@dendronhq/common-server";
import { SeedCLICommand, SeedCLICommandOpts } from "@dendronhq/dendron-cli";
import {
  DConfig,
  SeedInitMode,
  SeedService,
  SeedUtils,
} from "@dendronhq/engine-server";
import path from "path";
import { runEngineTestV5 } from "../../../engine";
import {
  checkDir,
  checkFile,
  checkNotInDir,
  checkVaults,
} from "../../../utils";
import { TestSeedUtils } from "../../../utils/seed";

export const runSeedCmd = ({
  cmd,
  id,
  ...opts
}: { cmd: SeedCommands } & Omit<SeedCLICommandOpts, "server">) => {
  const cli = new SeedCLICommand();
  return cli.execute({ cmd, id, ...opts, server: {} as any });
};

const BAD_SEED_ID = () => "dendron.no-exist";

async function createSeed({ engine }: { engine: DEngineClient }) {
  const tmp = tmpDir().name;
  const { registryFile, seedDict } = await TestSeedUtils.createSeedRegistry({
    engine,
    wsRoot: tmp,
  });
  const seedId = TestSeedUtils.defaultSeedId();
  return { registryFile, seedDict, seedId };
}

describe("remove", () => {
  const cmd = SeedCommands.REMOVE;
  test("error: nothing to remove", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot }) => {
        const id = BAD_SEED_ID();
        const resp = (await runSeedCmd({
          cmd,
          id,
          engine,
          wsRoot,
        })) as { error: DendronError };
        expect(resp).toMatchSnapshot();
        expect(resp.error.status).toEqual(ERROR_STATUS.DOES_NOT_EXIST);
      },
      {
        expect,
      }
    );
  });

  test("ok: remove non-initialized seed", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot, vaults }) => {
        // create seed
        const {
          registryFile,
          seedDict,
          seedId: id,
        } = await createSeed({ engine });

        // add seed to config;
        const seed = seedDict[id];
        const seedService = new SeedService({ wsRoot, registryFile });
        await seedService.addSeedMetadata({ seed, wsRoot });

        // remove seed
        await seedService.removeSeed({ id });

        await checkFile({
          fpath: path.join(wsRoot, "dendron.yml"),
          snapshot: true,
        });
        checkVaults(
          {
            wsRoot,
            vaults,
          },
          expect
        );
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });

  test("ok: remove initialized seed", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot, vaults }) => {
        // create seed
        const { registryFile, seedId: id } = await createSeed({ engine });

        // add seed to config;
        const seedService = new SeedService({ wsRoot, registryFile });
        await seedService.addSeed({ id });

        // remove seed
        await seedService.removeSeed({ id });

        const seedPath = SeedUtils.seed2Path({ wsRoot, id });
        await checkNotInDir(
          { fpath: path.dirname(seedPath), snapshot: true },
          id
        );
        expect(id).toMatchSnapshot();
        await checkFile({
          fpath: path.join(wsRoot, "dendron.yml"),
          snapshot: true,
        });
        checkVaults(
          {
            wsRoot,
            vaults,
          },
          expect
        );
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });
});

describe("add", () => {
  const cmd = SeedCommands.ADD;

  test("error: does not exist", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot }) => {
        const id = "dendron.no-exist";
        const resp = (await runSeedCmd({
          cmd,
          id,
          engine,
          wsRoot,
        })) as { error: DendronError };

        expect(resp).toMatchSnapshot();
        expect(resp.error.status).toEqual(ERROR_STATUS.DOES_NOT_EXIST);
      },
      {
        expect,
      }
    );
  });

  test("ok: exists", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot, vaults }) => {
        const tmp = tmpDir().name;
        const { registryFile } = await TestSeedUtils.createSeedRegistry({
          engine,
          wsRoot: tmp,
        });
        const id = TestSeedUtils.defaultSeedId();
        (await runSeedCmd({
          cmd,
          id,
          engine,
          wsRoot,
          registryFile,
        })) as { error: DendronError };
        // seed should be added
        await checkDir(
          { fpath: path.join(wsRoot, "seeds", id), snapshot: true },
          "dendron.yml",
          "seed.yml",
          "vault"
        );
        // seed.yml should not be present in the workspace
        await checkNotInDir(
          {
            fpath: wsRoot,
            snapshot: true,
          },
          "seed.yml"
        );
        await checkFile({
          fpath: path.join(wsRoot, "dendron.yml"),
          snapshot: true,
        });
        await checkFile(
          {
            fpath: path.join(wsRoot, "dendron.code-workspace"),
            snapshot: true,
          },
          `${id}/vault`
        );
        checkVaults(
          {
            wsRoot,
            vaults: [{ fsPath: "vault", seed: id, name: id } as DVault].concat(
              vaults
            ),
          },
          expect
        );
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });

  test("ok: seed with site", async () => {
    await runEngineTestV5(
      async ({ engine, wsRoot, vaults }) => {
        const tmp = tmpDir().name;
        const { registryFile } = await TestSeedUtils.createSeedRegistry({
          engine,
          wsRoot: tmp,
          modifySeed: (seed) => {
            seed.site = {
              url: "https://foo.com",
              index: "foo",
            };
            return seed;
          },
        });
        const id = TestSeedUtils.defaultSeedId();
        (await runSeedCmd({
          cmd,
          id,
          engine,
          wsRoot,
          registryFile,
        })) as { error: DendronError };
        // seed should be added
        await checkDir(
          { fpath: path.join(wsRoot, "seeds", id), snapshot: true },
          "dendron.yml",
          "seed.yml",
          "vault"
        );
        // seed.yml should not be present in the workspace
        await checkNotInDir(
          {
            fpath: wsRoot,
            snapshot: true,
          },
          "seed.yml"
        );
        await checkFile(
          {
            fpath: path.join(wsRoot, "dendron.yml"),
            snapshot: true,
          },
          "url: 'https://foo.com'",
          "index: foo"
        );
        await checkFile(
          {
            fpath: path.join(wsRoot, "dendron.code-workspace"),
            snapshot: true,
          },
          `${id}/vault`
        );
        checkVaults(
          {
            wsRoot,
            vaults: [{ fsPath: "vault", seed: id, name: id } as DVault].concat(
              vaults
            ),
          },
          expect
        );
      },
      {
        expect,
        addVSWorkspace: true,
      }
    );
  });
});

describe("init", () => {
  const cmd = SeedCommands.INIT;
  const seed: SeedConfig = {
    name: "foo",
    publisher: "dendron",
    description: "some foo",
    license: "CC",
    root: "vault",
    repository: {
      type: "git",
      url: "",
    },
  };
  describe("create workspace", () => {
    const mode = SeedInitMode.CREATE_WORKSPACE;
    test(`basic`, async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          await runSeedCmd({
            cmd,
            engine,
            wsRoot,
            mode,
            config: seed,
          });
          await checkDir(
            {
              fpath: wsRoot,
              snapshot: true,
            },
            "dendron.yml",
            "seed.yml",
            "dendron.code-workspace",
            "vault"
          );
          await checkFile({
            fpath: path.join(wsRoot, "dendron.yml"),
            snapshot: true,
          });
          await checkFile({
            fpath: path.join(wsRoot, "seed.yml"),
            snapshot: true,
          });
          await checkVaults(
            {
              wsRoot,
              vaults: [
                {
                  fsPath: "vault",
                },
              ],
            },
            expect
          );
          expect(DConfig.getOrCreate(wsRoot).seeds).toEqual(undefined);
        },
        {
          expect,
          preSetupHook: async () => {},
          vaults: [],
        }
      );
    });
  });

  describe("convert workspace", () => {
    const mode = SeedInitMode.CONVERT_WORKSPACE;

    const runInit = async (opts: { engine: DEngineClient; wsRoot: string }) => {
      const resp = await runSeedCmd({
        cmd,
        mode,
        config: seed,
        ...opts,
      });
      return resp?.error as DendronError | undefined;
    };

    test(`error: no vaults`, async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const error = await runInit({ engine, wsRoot });
          expect(error?.message).toContain(
            "workspace must have exactly one vault"
          );
          await checkDir(
            {
              fpath: wsRoot,
              snapshot: true,
            },
            "dendron.yml"
          );
          await checkFile({
            fpath: path.join(wsRoot, "dendron.yml"),
            snapshot: true,
          });
          expect(DConfig.getOrCreate(wsRoot).seeds).toEqual(undefined);
        },
        {
          expect,
          vaults: [],
        }
      );
    });

    test(`error: wsVault`, async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const error = await runInit({ engine, wsRoot });
          expect(error?.message).toEqual("workspace vaults not supported");
        },
        {
          expect,
          vaults: [],
          workspaces: [
            {
              name: "fooWorkspace",
              vaults: [{ fsPath: "vault" }],
              remote: { type: "git", url: "" },
            },
          ],
        }
      );
    });

    test(`ok: diff root name`, async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot }) => {
          const error = await runInit({ engine, wsRoot });
          expect(error).toBeUndefined();
          await checkDir(
            {
              fpath: wsRoot,
              snapshot: true,
            },
            "dendron.yml",
            "seed.yml"
          );
          await checkFile({
            fpath: path.join(wsRoot, "dendron.yml"),
            snapshot: true,
          });
          await checkFile(
            {
              fpath: path.join(wsRoot, "seed.yml"),
              snapshot: true,
            },
            "root: fooVault"
          );
        },
        {
          expect,
          vaults: [{ fsPath: "fooVault" }],
        }
      );
    });
  });
});
