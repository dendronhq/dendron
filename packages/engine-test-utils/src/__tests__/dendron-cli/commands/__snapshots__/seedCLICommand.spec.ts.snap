// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add error: does not exist 1`] = `
Object {
  "error": [Error: seed dendron.no-exist does not exist],
}
`;

exports[`add ok: exists 1`] = `".git .gitignore dendron.code-workspace dendron.yml seed.yml vault"`;

exports[`add ok: exists 2`] = `"dendron.code-workspace dendron.yml hooks seeds vault1 vault2 vault3"`;

exports[`add ok: exists 3`] = `
"version: 1
vaults:
    -
        fsPath: vault
        seed: dendron.foo
        name: dendron.foo
    -
        fsPath: vault1
    -
        fsPath: vault2
    -
        fsPath: vault3
        name: vaultThree
useFMTitle: true
useNoteTitleForLink: true
noAutoCreateOnDefinition: true
noLegacyNoteRef: true
noXVaultWikiLink: true
lookupConfirmVaultOnCreate: false
mermaid: true
useKatex: true
autoFoldFrontmatter: true
journal:
    dailyDomain: daily
    name: journal
    dateFormat: y.MM.dd
    addBehavior: childOfDomain
    firstDayOfWeek: 1
site:
    copyAssets: true
    siteHierarchies:
        - root
    siteRootDir: docs
    usePrettyRefs: true
    title: Dendron
    description: Personal knowledge space
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault1
            - vault2
            - vaultThree
            - dendron.foo
initializeRemoteVaults: true
seeds:
    dendron.foo: {}
"
`;

exports[`add ok: exists 4`] = `
"{
    \\"folders\\": [
        {
            \\"path\\": \\"seeds/dendron.foo/vault\\",
            \\"name\\": \\"dendron.foo\\"
        },
        {
            \\"path\\": \\"vault1\\"
        },
        {
            \\"path\\": \\"vault2\\"
        },
        {
            \\"path\\": \\"vault3\\",
            \\"name\\": \\"vaultThree\\"
        }
    ],
    \\"settings\\": {},
    \\"extensions\\": {}
}"
`;

exports[`add ok: seed with site 1`] = `".git .gitignore dendron.code-workspace dendron.yml seed.yml vault"`;

exports[`add ok: seed with site 2`] = `"dendron.code-workspace dendron.yml hooks seeds vault1 vault2 vault3"`;

exports[`add ok: seed with site 3`] = `
"version: 1
vaults:
    -
        fsPath: vault
        seed: dendron.foo
        name: dendron.foo
    -
        fsPath: vault1
    -
        fsPath: vault2
    -
        fsPath: vault3
        name: vaultThree
useFMTitle: true
useNoteTitleForLink: true
noAutoCreateOnDefinition: true
noLegacyNoteRef: true
noXVaultWikiLink: true
lookupConfirmVaultOnCreate: false
mermaid: true
useKatex: true
autoFoldFrontmatter: true
journal:
    dailyDomain: daily
    name: journal
    dateFormat: y.MM.dd
    addBehavior: childOfDomain
    firstDayOfWeek: 1
site:
    copyAssets: true
    siteHierarchies:
        - root
    siteRootDir: docs
    usePrettyRefs: true
    title: Dendron
    description: Personal knowledge space
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault1
            - vault2
            - vaultThree
            - dendron.foo
initializeRemoteVaults: true
seeds:
    dendron.foo:
        site:
            url: 'https://foo.com'
            index: foo
"
`;

exports[`add ok: seed with site 4`] = `
"{
    \\"folders\\": [
        {
            \\"path\\": \\"seeds/dendron.foo/vault\\",
            \\"name\\": \\"dendron.foo\\"
        },
        {
            \\"path\\": \\"vault1\\"
        },
        {
            \\"path\\": \\"vault2\\"
        },
        {
            \\"path\\": \\"vault3\\",
            \\"name\\": \\"vaultThree\\"
        }
    ],
    \\"settings\\": {},
    \\"extensions\\": {}
}"
`;

exports[`init convert workspace error: no vaults 1`] = `"dendron.yml hooks"`;

exports[`init convert workspace error: no vaults 2`] = `
"version: 1
vaults: []
useFMTitle: true
useNoteTitleForLink: true
noAutoCreateOnDefinition: true
noLegacyNoteRef: true
noXVaultWikiLink: true
lookupConfirmVaultOnCreate: false
mermaid: true
useKatex: true
autoFoldFrontmatter: true
journal:
    dailyDomain: daily
    name: journal
    dateFormat: y.MM.dd
    addBehavior: childOfDomain
    firstDayOfWeek: 1
site:
    copyAssets: true
    siteHierarchies:
        - root
    siteRootDir: docs
    usePrettyRefs: true
    title: Dendron
    description: Personal knowledge space
initializeRemoteVaults: true
"
`;

exports[`init convert workspace ok: diff root name 1`] = `"dendron.yml fooVault hooks seed.yml"`;

exports[`init convert workspace ok: diff root name 2`] = `
"version: 1
vaults:
    -
        fsPath: fooVault
useFMTitle: true
useNoteTitleForLink: true
noAutoCreateOnDefinition: true
noLegacyNoteRef: true
noXVaultWikiLink: true
lookupConfirmVaultOnCreate: false
mermaid: true
useKatex: true
autoFoldFrontmatter: true
journal:
    dailyDomain: daily
    name: journal
    dateFormat: y.MM.dd
    addBehavior: childOfDomain
    firstDayOfWeek: 1
site:
    copyAssets: true
    siteHierarchies:
        - root
    siteRootDir: docs
    usePrettyRefs: true
    title: Dendron
    description: Personal knowledge space
    duplicateNoteBehavior:
        action: useVault
        payload:
            - fooVault
initializeRemoteVaults: true
"
`;

exports[`init convert workspace ok: diff root name 3`] = `
"name: foo
publisher: dendron
description: some foo
license: CC
root: fooVault
repository:
    type: git
    url: ''
"
`;

exports[`init create workspace basic 1`] = `".gitignore dendron.code-workspace dendron.yml hooks seed.yml vault"`;

exports[`init create workspace basic 2`] = `
"version: 1
vaults:
    -
        fsPath: vault
useFMTitle: true
useNoteTitleForLink: true
noAutoCreateOnDefinition: true
noLegacyNoteRef: true
noXVaultWikiLink: true
lookupConfirmVaultOnCreate: false
mermaid: true
useKatex: true
autoFoldFrontmatter: true
journal:
    dailyDomain: daily
    name: journal
    dateFormat: y.MM.dd
    addBehavior: childOfDomain
    firstDayOfWeek: 1
site:
    copyAssets: true
    siteHierarchies:
        - root
    siteRootDir: docs
    usePrettyRefs: true
    title: Dendron
    description: Personal knowledge space
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault
initializeRemoteVaults: true
"
`;

exports[`init create workspace basic 3`] = `
"name: foo
publisher: dendron
description: some foo
license: CC
root: vault
repository:
    type: git
    url: ''
"
`;

exports[`remove error: nothing to remove 1`] = `
Object {
  "error": [Error: seed with id dendron.no-exist not in dendron.yml],
}
`;

exports[`remove ok: remove initialized seed 1`] = `""`;

exports[`remove ok: remove initialized seed 2`] = `"dendron.foo"`;

exports[`remove ok: remove initialized seed 3`] = `
"version: 1
vaults:
    -
        fsPath: vault1
    -
        fsPath: vault2
    -
        fsPath: vault3
        name: vaultThree
useFMTitle: true
useNoteTitleForLink: true
noAutoCreateOnDefinition: true
noLegacyNoteRef: true
noXVaultWikiLink: true
lookupConfirmVaultOnCreate: false
mermaid: true
useKatex: true
autoFoldFrontmatter: true
journal:
    dailyDomain: daily
    name: journal
    dateFormat: y.MM.dd
    addBehavior: childOfDomain
    firstDayOfWeek: 1
site:
    copyAssets: true
    siteHierarchies:
        - root
    siteRootDir: docs
    usePrettyRefs: true
    title: Dendron
    description: Personal knowledge space
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault1
            - vault2
            - vaultThree
            - dendron.foo
initializeRemoteVaults: true
seeds: {}
"
`;

exports[`remove ok: remove non-initialized seed 1`] = `
"version: 1
vaults:
    -
        fsPath: vault1
    -
        fsPath: vault2
    -
        fsPath: vault3
        name: vaultThree
useFMTitle: true
useNoteTitleForLink: true
noAutoCreateOnDefinition: true
noLegacyNoteRef: true
noXVaultWikiLink: true
lookupConfirmVaultOnCreate: false
mermaid: true
useKatex: true
autoFoldFrontmatter: true
journal:
    dailyDomain: daily
    name: journal
    dateFormat: y.MM.dd
    addBehavior: childOfDomain
    firstDayOfWeek: 1
site:
    copyAssets: true
    siteHierarchies:
        - root
    siteRootDir: docs
    usePrettyRefs: true
    title: Dendron
    description: Personal knowledge space
    duplicateNoteBehavior:
        action: useVault
        payload:
            - vault1
            - vault2
            - vaultThree
            - dendron.foo
initializeRemoteVaults: true
seeds: {}
"
`;
